src/qunit.cpp

```diff
@@ -38,19 +38,42 @@ QUnit::QUnit(QInterfaceEngine eng, bitLenInt qBitCount, bitCapInt initState, std
     }
 }
 
+void QUnit::CopyState(QUnitPtr orig)
+{
+    shards.clear();
+    SetQubitCount(orig->GetQubitCount());
+
+    /* Set up the shards to refer to the new unit. */
+    std::map<QInterfacePtr, QInterfacePtr> otherUnits;
+    for (auto otherShard :orig->shards) {
+        QEngineShard shard;
+        shard.mapped = otherShard.mapped;
+        if (otherUnits.find(otherShard.unit) == otherUnits.end()) {
+            otherUnits[otherShard.unit] = CreateQuantumInterface(engine, engine, otherShard.unit->GetQubitCount(), 0, rand_generator);
+        }
+        shard.unit = otherUnits[otherShard.unit];
+        shards.push_back(shard);
+    }
+
+    for (auto otherUnit : otherUnits) {
+        otherUnits[otherUnit.first]->CopyState(otherUnit.first);
+    }
+}
+
 void QUnit::CopyState(QInterfacePtr orig)
 {
-    QInterfacePtr unit = CreateQuantumInterface(engine, engine, 1, 0, rand_generator);
+    QInterfacePtr unit = CreateQuantumInterface(engine, engine, orig->GetQubitCount(), 0, rand_generator);
     unit->CopyState(orig);
 
     shards.clear();
     SetQubitCount(orig->GetQubitCount());
 
     /* Set up the shards to refer to the new unit. */
-    bitLenInt i = 0;
-    for (auto shard : shards) {
+    for (bitLenInt i = 0; i < (orig->GetQubitCount()); i++) {
+        QEngineShard shard;
         shard.unit = unit;
-        shard.mapped = i++;
+        shard.mapped = i;
+        shards.push_back(shard);
```

Comment:
shards[i].unit = unit; shards[i].mapped = i;

Reply:
See the above comment. Same deal; it covers units of different sizes, with probably little appreciable difference in speed. If the units are the same size, reallocation is likely already entirely avoided.